test/common/network/udp_listener_impl_test.cc

```diff
@@ -365,30 +396,37 @@ TEST_P(UdpListenerImplTest, SendData) {
 
   dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
 
-  Buffer::InstancePtr result_buffer(new Buffer::OwnedImpl());
-  const uint64_t bytes_to_read = 11;
+  constexpr uint64_t bytes_to_read = 11;
+  char recv_buf[bytes_to_read + 1];
   uint64_t bytes_read = 0;
   int retry = 0;
 
+  auto& os_sys_calls = Api::OsSysCallsSingleton::get();
+  sockaddr_storage peer_addr;
+  socklen_t addr_len = sizeof(sockaddr_storage);
   do {
-    Api::IoCallUint64Result result =
-        result_buffer->read(client_socket_->ioHandle(), bytes_to_read - bytes_read);
-
-    if (result.ok()) {
+    Api::SysCallSizeResult result = os_sys_calls.recvfrom(
+        client_socket_->ioHandle().fd(), recv_buf + bytes_read, bytes_to_read - bytes_read, 0,
+        reinterpret_cast<struct sockaddr*>(&peer_addr), &addr_len);
+    if (result.rc_ >= 0) {
       bytes_read += result.rc_;
-    } else if (retry == 10 || result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+      Address::InstanceConstSharedPtr peer_address =
+          Address::addressFromSockAddr(peer_addr, addr_len, false);
+      EXPECT_EQ(send_from_addr->asString(), peer_address->asString());
+    } else if (retry == 10 || result.errno_ != EAGAIN) {
       break;
     }
 
-    if (bytes_read == bytes_to_read) {
+    if (bytes_read >= bytes_to_read) {
       break;
     }
 
     retry++;
     ::usleep(10000);
   } while (true);
-
-  EXPECT_EQ(result_buffer->toString(), payload);
+  EXPECT_EQ(bytes_to_read, bytes_read);
```

Comment:
The loop above should be removed, the test only sends exactly one udp packet to the client, the client should receive it in exactly one `recvfrom` call.

Reply:
If it's not easy to change to a blocking recvfrom, I'm ok to keep the loop.  But since it's UDP, a recvfrom call should return all or nothing, so maybe remove the code where we try to "continue to read into the middle of recv_buf"?

Reply:
fair point. Done